### DISCUSSION TITLE
Decouple validation from on_change in input field

### DISCUSSION THREAD

**Discussion by Ch41r05** [zauberzeug/nicegui#2618] [Ideas / Feature Requests]
Hello everyone,

While trying to make a form with nicegui, I hit a wall during the field validation process.
My specific case is inherent to use the validate() method when the on blur event occurs, but this scenario could also be extended to other cases.

From digging a bit into nicegui internals I found

nicegui\elements\mixins\validation_element.py:48-50

```py
    def _handle_value_change(self, value: Any) -> None:
        super()._handle_value_change(value)
        self.validate()
```
So, from my undestanding, the validation and the on_change properties have a "silent" coupling which I don't know if it's a good way to link the two.

Would removing the self.valiate() from the method (hence using the parent's method) and letting the .on() event listeners manage validation be a bad idea?

As of now the way to do an on blur activation of the validate() method seems to be the following:

```py
ui.input(validation={"no value!": lambda e: bool(e)}).on("blur", lambda e: e.sender.validate())
```

Could this be a starting point to make validation triggers decoupled from other events?

Decoupling the validate() method from the on_change event would add the following benefits:

- Possibility to choose when the validate() method should be called, without additional unwanted calls
- Potential increase in performance if the validate() method needs to call external resources to check for validation
- In case of complex validation rules, limits the times that the submit button is enabled/disabled
- Easier aggregated validation of elements when the submit button of the form is pressed

These are just some scenarios that come to mind.
Most of all to have the freedom on the governance on when the validate() method is fired is obviously king here.

Please let me know your thoughts on this, since as of now the default behavior doesn't seem to be easily customizable.

**Comment by falkoschindler** [comment]
Thanks for the feature request, @Ch41r05!

We think the auto-validation is a handy feature in most cases.
But I just created PR #2661 to introduce an opt-out possibility.

**Reply by falkoschindler** [reply]
This feature just got released in [version 1.4.18](https://github.com/zauberzeug/nicegui/releases/tag/v1.4.18).
